sdks/java/core/src/main/java/org/apache/beam/sdk/io/TFRecordIO.java

```diff
@@ -0,0 +1,878 @@
+    }
+
+    private static WritableByteChannelFactory writableByteChannelFactory(
+        TFRecordIO.CompressionType compressionType) {
+      switch (compressionType) {
+        case AUTO:
+          throw new IllegalArgumentException("Unsupported compression type AUTO");
+        case NONE:
+          return CompressionType.UNCOMPRESSED;
+        case GZIP:
+          return CompressionType.GZIP;
+        case DEFLATE:
+          return CompressionType.DEFLATE;
+      }
+      return CompressionType.UNCOMPRESSED;
+    }
+
+    /**
+     * A {@link org.apache.beam.sdk.io.FileBasedSink.FileBasedWriteOperation
+     * FileBasedWriteOperation} for TFRecord files.
+     */
+    private static class TFRecordWriteOperation extends FileBasedWriteOperation<byte[]> {
+      private TFRecordWriteOperation(TFRecordSink sink) {
+        super(sink);
+      }
+
+      @Override
+      public FileBasedWriter<byte[]> createWriter(PipelineOptions options) throws Exception {
+        return new TFRecordWriter(this);
+      }
+    }
+
+    /**
+     * A {@link org.apache.beam.sdk.io.FileBasedSink.FileBasedWriter FileBasedWriter}
+     * for TFRecord files.
+     */
+    private static class TFRecordWriter extends FileBasedWriter<byte[]> {
+      private WritableByteChannel outChannel;
+      private TFRecordCodec codec;
+
+      private TFRecordWriter(FileBasedWriteOperation<byte[]> writeOperation) {
+        super(writeOperation);
+        this.mimeType = MimeTypes.BINARY;
+      }
+
+      @Override
+      protected void prepareWrite(WritableByteChannel channel) throws Exception {
+        this.outChannel = channel;
+        this.codec = new TFRecordCodec();
+      }
+
+      @Override
+      public void write(byte[] value) throws Exception {
+        codec.write(outChannel, value);
+      }
+    }
+  }
+
+  //////////////////////////////////////////////////////////////////////////////
+
+  private static class TFRecordCodec {
+    private static final int HEADER_LEN = (Long.SIZE + Integer.SIZE) / Byte.SIZE;
+    private static final int FOOTER_LEN = Integer.SIZE / Byte.SIZE;
+    private static HashFunction crc32c = Hashing.crc32c();
+
+    private ByteBuffer header = ByteBuffer.allocate(HEADER_LEN).order(ByteOrder.LITTLE_ENDIAN);
+    private ByteBuffer footer = ByteBuffer.allocate(FOOTER_LEN).order(ByteOrder.LITTLE_ENDIAN);
+
+    private int mask(int crc) {
+      return ((crc >>> 15) | (crc << 17)) + 0xa282ead8;
+    }
+
+    private int hashLong(long x) {
+      return mask(crc32c.hashLong(x).asInt());
+    }
+
+    private int hashBytes(byte[] x) {
+      return mask(crc32c.hashBytes(x).asInt());
+    }
+
+    public int recordLength(byte[] data) {
+      return HEADER_LEN + data.length + FOOTER_LEN;
+    }
+
+    public byte[] read(ReadableByteChannel inChannel) throws IOException {
+      header.clear();
+      if (inChannel.read(header) <= 0) {
+        return null;
+      }
+      header.rewind();
+      long length = header.getLong();
```

Comment:
As I'm not familiar with Java, I'm not sure if size of each type varies between systems (e.g. 32bit vs 64bit) or not.
If it varies, this should be handled carefully. If not (long=64bit, int=32bit always), never mind. 
Also please note that these are unsigned in C++.
https://github.com/tensorflow/tensorflow/blob/d24e64227653ab3cab7405c79ca166063a42f0ff/tensorflow/core/lib/io/record_reader.cc#L151
It should be equivalent.